test/RequestFactoryTest.php

```diff
@@ -2,34 +2,62 @@
 
 namespace Interop\Http\Factory;
 
-final class RequestFactoryTest extends RequestFactoryTestCase
-{
-    /**
-     * {@inheritdoc}
-     */
-    protected function createRequestFactory()
-    {
-        if (!defined('REQUEST_FACTORY')) {
-            $this->markTestSkipped('Request factory class name not provided');
-        }
-
-        $factoryClass = REQUEST_FACTORY;
-
-        return new $factoryClass();
-    }
-
-    /**
-     * {@inheritdoc}
-     */
-    protected function createUri($uri)
-    {
-        if (!defined('URI_FACTORY')) {
-            $this->markTestSkipped('URI factory class name not provided');
-        }
-
-        $factoryClass = URI_FACTORY;
-        $uriFactory = new $factoryClass();
-
-        return $uriFactory->createUri($uri);
-    }
+use PHPUnit\Framework\Attributes\DataProvider;
+use PHPUnit\Framework\TestCase;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\UriFactoryInterface;
+use function class_exists;
+use function defined;
+
+class RequestFactoryTest extends TestCase{
+
+	protected RequestFactoryInterface $requestFactory;
+	protected UriFactoryInterface     $uriFactory;
+
+	/**
+	 * @noinspection PhpUndefinedConstantInspection
+	 */
+	public function setUp():void{
+
+		if(!defined('REQUEST_FACTORY') || !class_exists(REQUEST_FACTORY)){
+			static::markTestSkipped('REQUEST_FACTORY class name not provided');
+		}
+
+		if(!defined('URI_FACTORY') || !class_exists(URI_FACTORY)){
+			static::markTestSkipped('URI_FACTORY class name not provided');
+		}
+
+		$this->requestFactory = new (REQUEST_FACTORY);
+		$this->uriFactory     = new (URI_FACTORY);
```

Comment:
This change is out of scope with supporting PHPUnit 10. Please revert it.

Reply:
The scope of this change is given in the OP.

Reply:
> random abstract classes and traits

These are not "random", they are design decisions that have downstream impacts. Changing this package's API has consequences.

Reply:
Ok, so I checked the most popular dependents for this library and all except one are using it as a stand-alone test suite without any changes. The Slim framework is the only one that extends the test classes and because of the "design decision" it has to unnecessarily re-implement several methods that are only found in the finalized classes here. Since you want to increment the major version you should take the opportunity to amend bad design decisions. What i can do is to rename the classes back to "*TestCase" so that libraries that extend them won't break - so even with out changes after an upgrade, their tests will still run and the old methods are ignored.

Reply:
There are two separate concerns that need to be separated:

1. PR #47 wasn't a complete switch to PHPUnit 10. I merged it and reduced the scope to be PHP 8.0+ and PHPUnit 10.0+. Obviously that was not quite sufficient, given that PHPUnit 10 requires PHP 8.1. So one change should be _just_ finalizing the PHPUnit 10.x upgrade, with as few code changes as possible.
2. This PR shows that we could make some code/API changes that would improve the test cases, and I am absolutely for that. But such changes would be represent version 3.x, because this package follows semantic versioning. Those changes need to be a separate PR that downstream users can review and have input on.

I am happy to review and merge fixes for both (1) and (2) but they need to be separate PRs.